File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/security/roles/service/UserRolesService.java
package com.github.bumblebee.security.roles.service;

import com.github.bumblebee.security.UserRole;
import com.github.bumblebee.security.roles.dao.UserRolesRepository;
import com.github.bumblebee.security.roles.domain.UserRoleRecord;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class UserRolesService {

    private final UserRolesRepository rolesRepository;
    private boolean rolesInitialized;

    @Autowired
    public UserRolesService(UserRolesRepository rolesRepository) {
        this.rolesRepository = rolesRepository;
    }

    @Transactional(readOnly = true)
    public boolean hasPrivilege(long userId, UserRole minimumRole) {

        return rolesRepository.findByUserId(userId).stream()
                .anyMatch(rec -> rec.getRole().rolePriority() >= minimumRole.rolePriority());
    }

    @Transactional(readOnly = true)
    public List<UserRole> getRoles(long userId) {
        return rolesRepository.findByUserId(userId).stream()
                .map(UserRoleRecord::getRole)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public Iterable<UserRoleRecord> getAllRoles() {
        return rolesRepository.findAll();
    }

    public void revokeRoles(long userId) {
        rolesRepository.deleteByUserId(userId);
    }

    public void revokeAllRoles() {
        rolesRepository.deleteAll();
    }

    public void setRoles(long userId, UserRole... roles) {

        revokeRoles(userId);

        Arrays.stream(roles)
                .map(role -> new UserRoleRecord(userId, role))
                .forEach(rolesRepository::save);
    }

    public boolean isRolesInitialized() {
        if (!rolesInitialized) {
            rolesInitialized = rolesRepository.count() > 0;
        }
        return rolesInitialized;
    }
}

File: telegram-bot-bumblebee-core/src/main/kotlin/com/github/bumblebee/security/roles/service/UserRolesService.kt
package com.github.bumblebee.security.roles.service

import com.github.bumblebee.security.UserRole
import com.github.bumblebee.security.roles.dao.UserRolesRepository
import com.github.bumblebee.security.roles.domain.UserRoleRecord
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class UserRolesService(val rolesRepository: UserRolesRepository) {
    private var rolesInitialized: Boolean = false

    @Transactional(readOnly = true)
    fun getAllRoles(): Iterable<UserRoleRecord> = rolesRepository.findAll()

    fun isRolesInitialized(): Boolean {
        if (!rolesInitialized) {
            rolesInitialized = rolesRepository.count() > 0
        }
        return rolesInitialized
    }

    @Transactional(readOnly = true)
    fun hasPrivilege(userId: Long, minimumRole: UserRole): Boolean {
        return rolesRepository.findByUserId(userId).any {
            it.role!!.rolePriority() >= minimumRole.rolePriority()
        }
    }

    @Transactional(readOnly = true)
    fun getRoles(userId: Long): List<UserRole> {
        return rolesRepository.findByUserId(userId).map { it.role!! }
    }

    fun revokeRoles(userId: Long) {
        rolesRepository.deleteByUserId(userId)
    }

    fun revokeAllRoles() {
        rolesRepository.deleteAll()
    }

    fun setRoles(userId: Long, vararg roles: UserRole) {

        revokeRoles(userId)

        Arrays.stream(roles)
                .map { role -> UserRoleRecord(userId, role) }
                .forEach({ rolesRepository.save(it) })
    }
}
